### PR TITLE
Async control & api update

### PR DESCRIPTION
**Pull request by GreenWizard2015**
I don't think there's any point in describing the code :)

The main issue is with compiling the unit tests. I've set up a native environment and separated the platform-dependent code, but for some reason, all files are being compiled during the build, which is causing problems.

### CODE REVIEW

**Comment by psmgeelen** [comment]
Awesome!

**Comment by psmgeelen** [comment]
https://docs.platformio.org/en/latest/advanced/unit-testing/index.html

**Comment by psmgeelen** [comment]
https://piolabs.com/blog/insights/unit-testing-part-1.html

**Comment by psmgeelen** [comment]
@omonrise has debugged most issues, but we might have an issue with the testing suite itself: https://docs.platformio.org/en/latest/advanced/unit-testing/frameworks/index.html It seems that not every platform is supported by each testing suite

**Comment by psmgeelen** [comment]
https://github.com/ThrowTheSwitch/Unity/blob/master/examples/unity_config.h

**Comment by psmgeelen** [comment]
And there something for local development called Arduino Mock?!? https://github.com/platformio/platformio-examples/tree/develop/unit-testing/arduino-mock

**Comment by omonrise** [comment]
tried to switch to unity testing framework aaand

In file included from .pio\libdeps\uno_r4_wifi\Unity\src\unity.h:21:0,
                 from .pio\libdeps\uno_r4_wifi\Unity\src\unity.c:7:
.pio\libdeps\uno_r4_wifi\Unity\src\unity_internals.h:11:10: fatal error: unity_config.h: No such file or directory

**********************************************************************
* Looking for unity_config.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:unity_config.h"
* Web  > https://registry.platformio.org/search?q=header:unity_config.h
*
**********************************************************************

 #include "unity_config.h"
          ^~~~~~~~~~~~~~~~
compilation terminated.
*** [.pio\build\uno_r4_wifi\lib062\Unity\unity.c.o] Error 1
======================================================================

Hi @GreenWizard2015  :)

**Comment by psmgeelen** [comment]
@GreenWizard2015 , so we put in some effort to make this work. Its not 100% functional yet. Could you review all the links that we have shared. Our learning from our session were:
* There were some issues with references that were called on as if they were pointers (arrows versus dots)
* gtest doesnt work for Arduino, so we migrated to unity
* we removed the local environment as we are not sure whether we need this right now
* We need to look into the platformio examples more to understand how this is all connected. A great resource is https://github.com/platformio/platformio-examples/tree/develop/unit-testing/arduino-mock

Would this be enough for you to look into it again and see whether you can solve it?

**Comment by GreenWizard2015** [comment]
@omonrise yeah, hi)

Well, everything didn't go quite as I wanted, so I redid everything) Now the project assembles properly, the tests work, and even their status is visible in VSCode.

I really don't like the idea of testing with a fake Android. To be more precise, I don't want to test absolutely everything this way. I believe it's better to abstract as much as possible from the hardware/platform. And only in extreme cases, test with hardware emulation.

File: .github/workflows/platformio.yaml [hunk 1]
```diff
@@ -20,6 +20,9 @@ jobs:
       - name: Install PlatformIO Core
         run: pip install --upgrade platformio
 
-      - name: Build PlatformIO Project
+      - name: Run tests on the native platform
+        run: platformio test -e native
+
+      - name: Build PlatformIO Project for Android
```
Comment (psmgeelen) [inline .github/workflows/platformio.yaml:27]: @GreenWizard2015  you probably mean Arduino instead of Android right?

**Comment by psmgeelen** [comment]
So I am fine with skipping the emulator stuff, :) I still have a few notes:
* Am I understanding correctly that we test the code on the native and build for Arduino? If so, this is something that I would strongly favour too, and would consider a good solution for our use-case. 
* A noticed a typo and made a comment
* Because we test only on native, we dont have to care for compatability of the framework, and therefore can use gtest. Is that a correct statement?

File: .github/workflows/platformio.yaml (hunk 1 above)
Comment (psmgeelen) [reply .github/workflows/platformio.yaml]: Yeah C++ does that to you :rofl:

**Comment by GreenWizard2015** [comment]
Yes, everything is as you described. 

I'm not sure if there would be issues due to FakeArduino, but my experience suggests that it's better to avoid such things. This way, the tests are more reliable, and the code is more thoughtful because there are strict limitations and you can't just lump everything together in a way that somehow works.

**Comment by psmgeelen** [comment]
I was cheeky and split out the CI, to have a clearer feedback of the CI-checks.

**Comment by psmgeelen** [comment]
Ow, and what would a test protocol look like with all the code changes? Specifically a question for @GreenWizard2015

**Comment by GreenWizard2015** [comment]
In general, its ready and I'm waiting for the feedback :-)

**Comment by psmgeelen** [comment]
As discussed,

test protocol:
1. trigger the pump and let it run
2. trigger the pump and test status
3. trigger the pump and interrupt

I will be testing it in real life today or tomorrow

**Comment by psmgeelen** [comment]
I can confirm that everything works:
1. Build works locally:
```
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [===       ]  27.9% (used 9128 bytes from 32768 bytes)
Flash: [========  ]  77.1% (used 202032 bytes from 262144 bytes)
Building .pio/build/uno_r4_wifi/firmware.bin
============================================================================ [SUCCESS] Took 0.99 seconds ============================================================================

Environment    Status    Duration
-------------  --------  ------------
uno_r4_wifi    SUCCESS   00:00:00.986
============================================================================ 1 succeeded in 00:00:00.986 ============================================================================
 *  Terminal will be reused by tasks, press any key to close it. 

```

Note that we are almost out of memory, so lets stop adding features ;)

2. Test pass locally:
```
Processing test_native in native environment
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Building...
Library Manager: Installing google/googletest @ ^1.12.1
Unpacking  [####################################]  100%
Library Manager: googletest@1.12.1 has been installed!
Testing...
WaterPumpScheduler.test_pump_stops_after_given_time     [PASSED]
WaterPumpScheduler.test_pump_is_periodically_forced_to_stop_after_given_time    [PASSED]
CommandProcessor.pour_tea_invalid_milliseconds  [PASSED]
CommandProcessor.pour_tea       [PASSED]
CommandProcessor.stop   [PASSED]
CommandProcessor.status [PASSED]
CommandProcessor.status_running [PASSED]
------------------------------------------------------------------- native:test_native [PASSED] Took 5.11 seconds -------------------------------------------------------------------

====================================================================================== SUMMARY ======================================================================================
Environment    Test         Status    Duration
-------------  -----------  --------  ------------
native         test_native  PASSED    00:00:05.112
===================================================================== 7 test cases: 7 succeeded in 00:00:05.112 =====================================================================
 *  Terminal will be reused by tasks, press any key to close it. 
```
3. The uploading/monitoring and connecting to WiFi works

```
Done in 0.001 seconds
Write 202032 bytes to flash (50 pages)

[                              ] 0% (0/50 pages)
[=                             ] 4% (2/50 pages)
[=                             ] 6% (3/50 pages)
[==                            ] 8% (4/50 pages)
[===                           ] 10% (5/50 pages)
[===                           ] 12% (6/50 pages)
[====                          ] 14% (7/50 pages)
[====                          ] 16% (8/50 pages)
[=====                         ] 18% (9/50 pages)
[======                        ] 20% (10/50 pages)
[======                        ] 22% (11/50 pages)
[=======                       ] 24% (12/50 pages)
[=======                       ] 26% (13/50 pages)
[========                      ] 28% (14/50 pages)
[=========                     ] 30% (15/50 pages)
[=========                     ] 32% (16/50 pages)
[==========                    ] 34% (17/50 pages)
[==========                    ] 36% (18/50 pages)
[===========                   ] 38% (19/50 pages)
[============                  ] 40% (20/50 pages)
[============                  ] 42% (21/50 pages)
[=============                 ] 44% (22/50 pages)
[=============                 ] 46% (23/50 pages)
[==============                ] 48% (24/50 pages)
[===============               ] 50% (25/50 pages)
[===============               ] 52% (26/50 pages)
[================              ] 54% (27/50 pages)
[================              ] 56% (28/50 pages)
[=================             ] 58% (29/50 pages)
[==================            ] 60% (30/50 pages)
[==================            ] 62% (31/50 pages)
[===================           ] 64% (32/50 pages)
[===================           ] 66% (33/50 pages)
[====================          ] 68% (34/50 pages)
[=====================         ] 70% (35/50 pages)
[=====================         ] 72% (36/50 pages)
[======================        ] 74% (37/50 pages)
[======================        ] 76% (38/50 pages)
[=======================       ] 78% (39/50 pages)
[========================      ] 80% (40/50 pages)
[========================      ] 82% (41/50 pages)
[=========================     ] 84% (42/50 pages)
[=========================     ] 86% (43/50 pages)
[==========================    ] 88% (44/50 pages)
[===========================   ] 90% (45/50 pages)
[===========================   ] 92% (46/50 pages)
[============================  ] 94% (47/50 pages)
[============================  ] 96% (48/50 pages)
[============================= ] 98% (49/50 pages)
[==============================] 100% (50/50 pages)
Done in 12.519 seconds
--- Terminal on /dev/ttyACM0 | 9600 8-N-1
--- Available filters and text transformations: colorize, debug, default, direct, hexlify, log2file, nocontrol, printable, send_on_enter, time
--- More details at https://bit.ly/pio-monitor-filters
--- Quit: Ctrl+C | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H
Latest available version: 0.3.0
Please upgrade your firmware.
Connecting to FamGeelen2.4Ghz
Atempt to connect: 1
.
Connection status: 3

Connected. IP: 192.168.178.60
SSID: FamGeelen2.4Ghz
BSSID: DC:39:6F:B8:B8:E3
signal strength (RSSI): -59
Encryption Type: 2
----------------------------------------------
```

4. And the API itself works, too. I will be releasing a Video to showcase how to test in practise. @GreenWizard2015 , feel free to merge!

**Review by psmgeelen — APPROVED** [review]

**Comment by GreenWizard2015** [comment]
just in case, don't merge till we done full testing of this and UI branch.

**Comment by psmgeelen** [comment]
All testing done

- Build successful
- API tests successful

**Comment by psmgeelen** [comment]
Lets merge :)

**Pull request merged**